### PR TITLE
Support signing typed data on Trezor devices

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/41864

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
`Sign Typed Data V3` and `Sign Typed Data V4`  on https://metamask.github.io/test-dapp/
should work for Trezor accounts with firmware newer than `T1B1`

### CODE REVIEW

File: components/brave_wallet/common/eth_request_helper.cc [hunk 1]
```diff
@@ -240,12 +241,16 @@ bool GetEthJsonRequestInfo(const std::string& json,
...
-    const auto* found_params = response_dict->FindListByDottedPath(kParams);
-    if (!found_params) {
-      return false;
+  if (params_list) {
+    params_list->clear();
+    if (auto* found_params = response_dict->FindByDottedPath(kParams)) {
+      if (found_params->is_list()) {
+        *params_list = std::move(found_params->GetList());
+      } else if (!found_params->is_dict()) {
+        // dict is a valid type of params by we don't support it.
```
Comment (yrliou) [inline components/brave_wallet/common/eth_request_helper.cc:250]: typo: s/by/but

File: components/brave_wallet/browser/ethereum_provider_impl.cc [hunk 2]
```diff
@@ -108,22 +108,14 @@ void RejectMismatchError(base::Value id,
...
-  if (ParseEthSignTypedDataParams(normalized_json_request, &address, &message,
-                                  &domain, EthSignTypedDataHelper::Version::kV4,
-                                  &domain_hash_out, &primary_hash_out, &meta) ||
-      ParseEthSignTypedDataParams(normalized_json_request, &address, &message,
-                                  &domain, EthSignTypedDataHelper::Version::kV3,
-                                  &domain_hash_out, &primary_hash_out, &meta)) {
-    return true;
+bool IsTypedDataStructure(const base::Value::List& params_list) {
+  if (ParseEthSignTypedDataAddress(params_list)) {
+    if (ParseEthSignTypedDataParams(params_list,
```
Comment (yrliou) [inline components/brave_wallet/browser/ethereum_provider_impl.cc:113]: Would it be better to include address in `EthSignTypedData` and just parse it in `ParseEthSignTypedDataParams`?
Will need to revise `SignTypedMessage`, but seems easier to read to me and we don't need to do address null checks outside of the parser function.

Comment (supermassive) [reply components/brave_wallet/browser/ethereum_provider_impl.cc]: Added https://github.com/brave/brave-core/pull/26165/commits/77caacd564fe1cc0b8620d9d090be0da54030f00

**Review by yrliou — APPROVED** [review]

**Review by onyb — APPROVED** [review]
frontend++

Please add QA steps to the PR or linked issue.

**Pull request merged**